waspc/src/Wasp/AppSpec/Valid.hs

```diff
@@ -50,52 +51,71 @@ validateExactlyOneAppExists spec =
-                  _ ->
-                    [ GenericValidationError
-                        "Expected an Entity referenced by app.auth.userEntity to have field 'password' of type 'string'."
-                    ]
+         in concatMap
+              (validateField "app.auth.userEntity" userEntityFields)
+              [ ("username", Entity.Field.FieldTypeScalar Entity.Field.String, "String"),
+                ("password", Entity.Field.FieldTypeScalar Entity.Field.String, "String")
               ]
 
+validateExternalAuthAssociationEntityHasCorrectFieldsIfExternalAuthIsUsed :: AppSpec -> [ValidationError]
+validateExternalAuthAssociationEntityHasCorrectFieldsIfExternalAuthIsUsed spec = case App.auth (snd $ getApp spec) of
+  Nothing -> []
+  Just auth ->
+    if not $ Auth.isExternalAuthEnabled auth
+      then []
+      else case Auth.externalAuthAssociationEntity auth of
+        Nothing -> [GenericValidationError "Expected an Entity referenced by app.auth.externalAuthAssociationEntity to exist when using a social login method."]
+        Just externalAuthAssociationEntityRef ->
+          let (userEntityName, userEntity) = AS.resolveRef spec (Auth.userEntity auth)
+              userEntityFields = Entity.getFields userEntity
+              (externalAuthAssociationEntityName, externalAuthAssociationEntity) = AS.resolveRef spec externalAuthAssociationEntityRef
+              externalAuthAssociationEntityFields = Entity.getFields externalAuthAssociationEntity
+           in concatMap
+                (validateField "app.auth.externalAuthAssociationEntity" externalAuthAssociationEntityFields)
+                [ ("provider", Entity.Field.FieldTypeScalar Entity.Field.String, "String"),
+                  ("providerId", Entity.Field.FieldTypeScalar Entity.Field.String, "String"),
+                  ("user", Entity.Field.FieldTypeScalar (Entity.Field.UserType userEntityName), userEntityName),
+                  ("userId", Entity.Field.FieldTypeScalar Entity.Field.Int, "Int")
+                ]
+                ++ concatMap
+                  (validateField "app.auth.userEntity" userEntityFields)
+                  [ ("externalAuthAssociations", Entity.Field.FieldTypeComposite $ Entity.Field.List $ Entity.Field.UserType externalAuthAssociationEntityName, externalAuthAssociationEntityName ++ "[]")
+                  ]
+
+validateField :: String -> [Entity.Field.Field] -> (String, Entity.Field.FieldType, String) -> [ValidationError]
+validateField entityName entityFields (fieldName, fieldType, fieldTypeName) =
+  let maybeField = find ((== fieldName) . Entity.Field.fieldName) entityFields
+   in case maybeField of
+        Just providerField
+          | Entity.Field.fieldType providerField == fieldType -> []
+        _ ->
+          [ GenericValidationError $
+              "Expected an Entity referenced by " ++ entityName ++ " to have field '" ++ fieldName ++ "' of type '" ++ fieldTypeName ++ "'."
```

Comment:
Instead of receiving field type name, can you just do `show fieldType`?

Reply:
It produces a fully qualified name, and we also don't have a great way of handling arrays. In this way, I figured it was easiest to just show what the user-facing Prisma type would be instead of munging the internal type names.